support multi BLOCK_M for flex-decoding

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170343

Flex-decoding previously is only active in VLLM in very narrow cases. We use flex-decoding if the workload can be processed with a single BLOCK_M: i.e. `BLOCK_M >= seq_len_q * G`, where G is the ratio between Q head and KV head. We use BLOCK_M as 16 in the vllm integration.

Take llama3 8B as an example, G is 4 here. That means we use flex-decoding only if `seq_len_q` <= 4. This is very restrictive and make flex-decoding being skipped most of the time.

The PR enhance the flex-decoding kernel to work for multi-BLOCK_M.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

shunting314 committed:
support multi BLOCK_M for flex-decoding

[ghstack-poisoned]

shunting314 committed:
Update on "support multi BLOCK_M for flex-decoding"


Flex-decoding previously is only active in VLLM in very narrow cases. We use flex-decoding if the workload can be processed with a single BLOCK_M: i.e. `BLOCK_M >= seq_len_q * G`, where G is the ratio between Q head and KV head. We use BLOCK_M as 16 in the vllm integration.

Take llama3 8B as an example, G is 4 here. That means we use flex-decoding only if `seq_len_q` <= 4. This is very restrictive and make flex-decoding being skipped most of the time.

The PR enhance the flex-decoding kernel to work for multi-BLOCK_M.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

drisspg commented on torch/_inductor/kernel/flex/flex_decoding.py:
```diff
@@ -88,35 +78,35 @@ def _use_flex_decoding(query, kv_indices, value, kernel_options, enable_gqa) ->
-    return (batch_size * kv_heads, meta["SPLIT_KV"], 1)
+    BLOCK_M = meta["BLOCK_M"]
+    num_block_m = ceildiv(seq_len_q * gqa_group_size, BLOCK_M)
+    return (batch_size * kv_heads * num_block_m, meta["SPLIT_KV"], 1)
```
nit: why dont we just put block_m into its own dim: https://github.com/pytorch/pytorch/blob/a2d11eb06972cbfa3fb0d23595e89d7585618e57/torch/_inductor/kernel/flex/flex_attention.py#L75

drisspg reviewed (approved):
One nit, looks great!

shunting314 committed:
Update on "support multi BLOCK_M for flex-decoding"


Flex-decoding previously is only active in VLLM in very narrow cases. We use flex-decoding if the workload can be processed with a single BLOCK_M: i.e. `BLOCK_M >= seq_len_q * G`, where G is the ratio between Q head and KV head. We use BLOCK_M as 16 in the vllm integration.

Take llama3 8B as an example, G is 4 here. That means we use flex-decoding only if `seq_len_q` <= 4. This is very restrictive and make flex-decoding being skipped most of the time.

The PR enhance the flex-decoding kernel to work for multi-BLOCK_M.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge -i

shunting314 committed:
Update on "support multi BLOCK_M for flex-decoding"


Flex-decoding previously is only active in VLLM in very narrow cases. We use flex-decoding if the workload can be processed with a single BLOCK_M: i.e. `BLOCK_M >= seq_len_q * G`, where G is the ratio between Q head and KV head. We use BLOCK_M as 16 in the vllm integration.

Take llama3 8B as an example, G is 4 here. That means we use flex-decoding only if `seq_len_q` <= 4. This is very restrictive and make flex-decoding being skipped most of the time.

The PR enhance the flex-decoding kernel to work for multi-BLOCK_M.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]
